src/main/java/io/opentracing/contrib/global/GlobalTracer.java

```diff
@@ -58,23 +81,44 @@
      * This can be either an {@link #register(Tracer) explicitly registered} or
      * the {@link #loadSingleSpiImplementation() automatically resolved} Tracer
      * (or <code>null</code> during initialization).
+     * <p>
+     * Access regulated by {@link #LOCK}.
      */
-    private final AtomicReference<Tracer> globalTracer = new AtomicReference<Tracer>();
+    private static Tracer globalTracer = null;
+    private static final ReadWriteLock LOCK = new ReentrantReadWriteLock();
 
     private GlobalTracer() {
     }
 
-    private Tracer lazyTracer() {
-        Tracer tracer = globalTracer.get();
-        if (tracer == null) {
-            final Tracer resolved = loadSingleSpiImplementation();
-            while (tracer == null && resolved != null) { // handle rare race condition
-                globalTracer.compareAndSet(null, resolved);
-                tracer = globalTracer.get();
+    // acquires read lock and returns the globalTracer
+    private static Tracer globalTracer() {
+        LOCK.readLock().lock();
+        try {
+            return globalTracer;
+        } finally {
+            LOCK.readLock().unlock();
+        }
+    }
+
+    // lazily loading of the globalTracer
+    private static Tracer lazyTracer() {
+        LOCK.readLock().lock();
```

Comment:
@wu-sheng I may be a little overly paranoid.
I tested your suggestion with `ReentrantLock` and `volatile` variable and it seems to work just fine.